Blish HUD/Controls/TextInputBase.cs

```diff
@@ -0,0 +1,718 @@
+                    case OperationType.Insert:
+                        redoStack.MakeDelete(_text, record.Index, record.Length);
+                        DeleteChars(record.Index, record.Length);
+                        UserSetCursorIndex(record.Index);
+                        break;
+                    case OperationType.Delete:
+                        if (InsertChars(record.Index, record.Data, out int length)) {
+                            redoStack.MakeInsert(record.Index, length);
+                            UserSetCursorIndex(record.Index + length);
+                        }
+
+                        break;
+                    case OperationType.Replace:
+                        redoStack.MakeReplace(_text, record.Index, record.Length, record.Data.Length);
+                        DeleteChars(record.Index, record.Length);
+                        InsertChars(record.Index, record.Data, out _);
+                        break;
+                }
+            } finally {
+                _suppressRedoStackReset = false;
+            }
+
+            ResetSelection();
+        }
+
+        protected void UserSetCursorIndex(int newIndex) {
+            if (newIndex > _text.Length) {
+                newIndex = _text.Length;
+            }
+
+            if (newIndex < 0) {
+                newIndex = 0;
+            }
+
+            this.CursorIndex = newIndex;
+        }
+
+        protected void ResetSelection() {
+            this.SelectionStart = _selectionEnd = _cursorIndex;
+        }
+
+        protected void UpdateSelection() {
+            this.SelectionEnd = _cursorIndex;
+        }
+
+        protected void UpdateSelectionIfShiftDown() {
+            if (this.IsShiftDown) {
+                UpdateSelection();
+            } else {
+                ResetSelection();
+            }
+        }
+
+        protected abstract void MoveLine(int delta);
+
+        protected void SelectAll() {
+            this.SelectionStart = 0;
+            this.SelectionEnd   = _text.Length;
+        }
+
+        protected float MeasureStringWidth(string text) {
+            if (string.IsNullOrEmpty(text)) return 0;
+
+            var lastGlyph = _font.GetGlyphs(text).Last();
+
+            return lastGlyph.Position.X + (lastGlyph.FontRegion?.Width ?? 0);
+        }
+
+        private string ProcessText(string value) {
+            value = value?.Replace("\r", string.Empty);
+
+            if (!_multiline) {
+                value = value?.Replace("\n", string.Empty);
+            }
+
+            if (value?.Length > _maxLength) {
+                value = value.Substring(0, _maxLength);
+            }
+
+            return value;
+        }
+
+        protected bool SetText(string value, bool byUser) {
+            string prevText = _text;
+
+            value = ProcessText(value);
+
+            if (!SetProperty(ref _text, value)) return false;
+
+            // TODO: Update formatted text?
+
+            if (!byUser) {
```

Comment:
You could put that into the setter and use `SetText` inside this class w/o the boolean

Reply:
This is maintaining the convention used in Myra:
https://github.com/rds1983/Myra/blob/a9dbf7a1ceedc19f9e416c754eaf38e89a89a746/src/Myra/Graphics2D/UI/TextBox.cs#L860-L867

Also doesn't matter as a whole and letting `SetText` accept the bool lets implementing classes provide it as well.

Reply:
Implementing classes can still use `SetText` and the setter like the base class, so it won't matter there. But i cannot say anything against the first argument.